packages/react/src/hooks.ts

```diff
@@ -0,0 +1,77 @@
+import type {
+  Client,
+  Space,
+  Account,
+  ServiceConfig
+} from '@w3ui/core'
+
+import { useState, useEffect, useCallback } from 'react'
+import { STORE_SAVE_EVENT, createClient } from '@w3ui/core'
+
+export type DatamodelProps = ServiceConfig
+
+export interface Datamodel {
+  client?: Client
+  accounts: Account[]
+  spaces: Space[]
+  logout: () => Promise<void>
+}
+
+export function useDatamodel ({ servicePrincipal, connection }: DatamodelProps): Datamodel {
+  const [client, setClient] = useState<Client>()
+  const [events, setEvents] = useState<EventTarget>()
+  const [accounts, setAccounts] = useState<Account[]>([])
+  const [spaces, setSpaces] = useState<Space[]>([])
+
+  // update this function any time servicePrincipal or connection change
+  const setupClient = useCallback(
+    async (): Promise<void> => {
+      const { client, events } = await createClient({ servicePrincipal, connection })
+      setClient(client)
+      setEvents(events)
+      setAccounts(Object.values(client.accounts()))
+      setSpaces(client.spaces())
+    },
+    [servicePrincipal, connection]
+  )
+
+  // run setupClient once each time it changes
+  useEffect(
+    () => {
+      void setupClient()
+    },
+    [setupClient]
+  )
+
+  // set up event listeners to refresh accounts and spaces when
+  // the store:save event from @w3ui/core happens
+  useEffect(() => {
+    if ((client === undefined) || (events === undefined)) return
+
+    const handleStoreSave: () => void = () => {
+      setAccounts(Object.values(client.accounts()))
+      setSpaces(client.spaces())
+    }
+
+    events.addEventListener(STORE_SAVE_EVENT, handleStoreSave)
+    return () => {
+      events?.removeEventListener(STORE_SAVE_EVENT, handleStoreSave)
+    }
+  }, [client, events])
+
+  const logout = async (): Promise<void> => {
```

Comment:
personally I would recommend a `React.useCallback` here but i don't think it is needed for any functional reason. It might make the caller's component re-render slightly less but that can always be later. (we discussed this sync and you had reaons not to which are fine)